ports/cheriot-rtos/machine_spi.c

```diff
@@ -0,0 +1,300 @@
+    };
+
+    machine_spi_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args),
+        allowed_args, args);
+
+    uint32_t baudrate = args[ARG_baudrate].u_int;
+    uint8_t polarity = args[ARG_polarity].u_int;
+    uint8_t phase = args[ARG_phase].u_int;
+    uint8_t firstbit = args[ARG_firstbit].u_int;
+
+    uint32_t config = self->spi_block->configuration;
+
+    if (baudrate > 0) {
+        config &= ~ConfigurationHalfClockPeriodMask;
+        config |= convert_baudrate(baudrate);
+    }
+
+    if (polarity >= 0) {
+        config &= ~ConfigurationClockPolarity;
+        config |= (polarity ? ConfigurationClockPolarity : 0);
+    }
+
+    if (phase >= 0) {
+        config &= ~ConfigurationClockPhase;
+        config |= (phase ? ConfigurationClockPhase : 0);
+    }
+
+    if (firstbit >= 0) {
+        config &= ~ConfigurationMSBFirst;
+        config |= (firstbit ? ConfigurationMSBFirst : 0);
+    }
+
+    self->spi_block->configuration = config;
+}
+
+static void machine_spi_transfer(mp_obj_base_t *obj, size_t len,
+    const uint8_t *src, uint8_t *dest) {
+    if (len >= 0x7ff) {
+        mp_raise_ValueError(
+            MP_ERROR_TEXT("cannot transfer more than 2047 bytes at a time"));
+    }
+
+    machine_spi_obj_t *self = MP_OBJ_TO_PTR(obj);
+
+    volatile spi_block_t *spi_block = self->spi_block;
+
+    wait_idle(spi_block);
+
+    bool rcv = false;
+    bool trx = false;
+    uint32_t control = 0;
+
+    uint32_t tx_count = len;
+    uint32_t rx_count = len;
+
+    if (src) {
+        trx = true;
+        control = ControlTransmitEnable;
+        tx_count = 0;
+    }
+    if (dest) {
+        rcv = true;
+        control |= ControlReceiveEnable;
+        rx_count = 0;
+    }
+
+    spi_block->control = control;
+    spi_block->start = len & StartByteCountMask;
+
+    while (tx_count < len || rx_count < len) {
+        uint32_t status = spi_block->status;
+        if (trx && (status & StatusTxFifoLevel) < 64 && tx_count < len) {
+            spi_block->transmitFifo = src[tx_count];
+            printf("tx >>> %02x -- %08x\n", src[tx_count],
+                (uint32_t)&(spi_block->transmitFifo));
+            tx_count++;
+        } else if (rcv && (status & StatusRxFifoLevel) > 0 && rx_count < len) {
+            dest[rx_count] = spi_block->receiveFifo;
+            printf("rx >>> %02x -- %08x\n", dest[rx_count],
+                (uint32_t)&(spi_block->receiveFifo));
+
+            rx_count++;
+        }
+    }
+}
+
+static const mp_machine_spi_p_t machine_spi_protocol = {
+    .init = machine_spi_init,
+    .transfer = machine_spi_transfer,
+};
+
+MP_DEFINE_CONST_OBJ_TYPE(machine_spi_type, MP_QSTR_SPI, MP_TYPE_FLAG_NONE,
+    make_new, machine_spi_make_new, call, print,
+    machine_spi_print, protocol, &machine_spi_protocol,
+    locals_dict, &mp_machine_spi_locals_dict);
```

Comment:
This block yields a compiler error because of the `call, ` not being followed by a valid call function.